### DISCUSSION TITLE
how to run aio image using podman (docker replacement)

### DISCUSSION THREAD

**Discussion by ncolin1999** [nextcloud/all-in-one#5090] [Questions] [answered]
Some one can help!!

**Comment by szaimen** [comment, marked as answer]
Hi, see https://github.com/nextcloud/all-in-one?tab=readme-ov-file#can-i-run-this-with-podman-instead-of-docker

**Reply by aanno** [reply]
While @szaimen gives the correct answer: This is _not_ supported but you can try the [manual-install](https://github.com/nextcloud/all-in-one/blob/main/manual-install) or look at the [Rootless Podman Quadlet #3487 discussion](https://github.com/nextcloud/all-in-one/discussions/3487), I would like to share my personal (i.e. completly biased) experience.

I'm running Nextcloud AIO on [FCOS](https://fedoraproject.org/coreos/) with rootless podman (but with socket API enabled). _This basically works (!)_ **but** there are some issues. In short: No [self tests](https://docs.nextcloud.com/server/30/admin_manual/configuration_server/security_setup_warnings.html), no talk, no collabora, etc.

You could see what I'm doing on _my own branch of all-in-one_ called 'aanno/aio-on-fcos-caddy-host-mode' [here](https://github.com/aanno/all-in-one/tree/aanno/aio-on-fcos-caddy-host-mode/aio-on-fcos).

The main problem running this setup is that the 'self reference' is broken. For example, this is `/etc/hosts` of running `nextcloud-aio-nextcloud` container:

```bash
$ podman exec -it nextcloud-aio-nextcloud sh
/var/www/html # cat /etc/hosts
127.0.0.1       localhost
::1     localhost
169.254.1.2     host.containers.internal host.docker.internal
10.89.1.9       9e42487a07a1 nextcloud-aio-nextcloud
```

But it should be more like this:

```bash
# cat /etc/hosts
127.0.0.1       localhost
::1     localhost
169.254.1.2     host.containers.internal host.docker.internal <your-nextcloud-aio-fq-domain>
10.89.1.9       9e42487a07a1 nextcloud-aio-nextcloud
```

While it is possible to fix that for `nextcloud-aio-nextcloud` (and after this the self tests work as expected), _the same fix does not work for many of the other containers_ as many of these have `/etc/hosts` read-only.

**Reply by eriksjolund** [reply]
> The main problem running this setup is that the 'self reference' is broken. 

Have you tried adding `--add-host your-nextcloud-aio-fq-domain:host-gateway` ?
This example requires Podman 5.3.0 or later:

```
$ podman run --rm --add-host your-nextcloud-aio-fq-domain:host-gateway alpine cat /etc/hosts | grep 169.254.1.2
169.254.1.2	your-nextcloud-aio-fq-domain
169.254.1.2	host.containers.internal host.docker.internal
$ podman --version
podman version 5.3.0-rc1
$ 
```
I wrote a little bit more about `--add-host exaxmple.com:host-gateway` in [example: connect to host's main network interface using pasta](https://github.com/eriksjolund/podman-networking-docs?tab=readme-ov-file#example-connect-to-hosts-main-network-interface-using-pasta)

**Reply by aanno** [reply]
Dear @eriksjolund ,

this would be an idea if I have used the [manual-install](https://github.com/nextcloud/all-in-one/blob/main/manual-install) (which I haven't). It is of course not feasible with normal/automatic mode that starts containers. Maybe it could be fixed this way in the 'control' container, though.

**Reply by apparle** [reply]
> > Have you tried adding --add-host your-nextcloud-aio-fq-domain:host-gateway ?

> ...
> this would be an idea if I have used the [manual-install](https://github.com/nextcloud/all-in-one/blob/main/manual-install) (which I haven't). It is of course not feasible with normal/automatic mode that starts containers. Maybe it could be fixed this way in the 'control' container, though.

This seems like a small thing to fix in `DockerActionManager :: CreateContainer` by adding:
```
if ($container->GetIdentifier() != 'nextcloud-aio-domaincheck') {
        $requestBody['HostConfig']['ExtraHosts'] = [ $this->configurationManager->GetDomain() . ":host-gateway"] ;
}
```
Maybe it could be guarded under a non-default env variable or even a dynamically detected if container runtime is podman instead of docker, to avoid adding any configuration variables. @szaimen what do you think?

@aanno I'm curious if this would fix your relatively complicated coredns based set up. I don't have podman v5.3.0 on my server yet, but you can probably try it.

**Reply by aanno** [reply]
Dear @apparle ,

I made a branch https://github.com/aanno/all-in-one/tree/aanno/5090-add-host that incorporates the suggested change. However, I'm a bit lost now. My first idea was to build the image (let's say docker.io/aanno/all-in-one:latest) on my own and then test it without the coredns stuff.

However, how to build the all-in-one image in the first place?

**Reply by apparle** [reply]
@aanno Well, I've found testing local edits to this repo quite difficult myself. But I'm trying to improve that on https://github.com/nextcloud/all-in-one/pull/5568 . I'd recommend you also pull in that PR's changes into your copy and follow my first and last comment to do local testing.

**Reply by rriemann** [reply]
My alternative approach is to have a specific unix user and configure for this one the network with slirp4netns. This does not require patching.

more: https://github.com/nextcloud/all-in-one/discussions/3487#discussioncomment-12971567

**Comment by aanno** [comment]
Hello everyone,

I like to announce that I was finally able to circumvent the 'self reference' problem. _Hence, I'm now running a full-blow and full-working all-in-one instance on **podman**_ (on [FCOS](https://fedoraproject.org/coreos/)). It is stable (and runs for weeks now) _but it is officially unsupported_.

You could see what I'm doing on my own branch of all-in-one called 'aanno/aio-on-fcos-coredns-wildcard' [here](https://github.com/aanno/all-in-one/tree/aanno/aio-on-fcos-coredns-wildcard/aio-on-fcos). (Look into the `aio-on-fcos` folder.)

There are 2 ideas that helped me to overcome the 'self reference' problem.

1. Use 2 external networks configured appropriate:
    ```bash
   podman network create --subnet 10.89.57.0/24 --gateway 10.89.57.1 --dns 10.89.57.4 --driver bridge --ipv6 --subnet fd49:dc34:d0fe:ef6b:cafe::/80 --gateway fd49:dc34:d0fe:ef6b:cafe::1 --dns fd49:dc34:d0fe:ef6b:cafe::4 nextcloud-aio
   podman network create --subnet 10.89.58.0/24 --gateway 10.89.58.1 --dns 10.89.58.4 --driver bridge --ipv6 --subnet fd49:dc34:d0fe:ef6b:beaf::/80 --gateway fd49:dc34:d0fe:ef6b:beaf::1 --dns fd49:dc34:d0fe:ef6b:beaf::4 nextcloud_frontend
    ```
2. Use [coredns](https://coredns.io/) (with some configuration) to resolve AIO_DOMAIN properly.

In addition, my solution has the following properties:

1. ipv6 support
2. Caddy used as reverse proxy.
3. The same caddy is used to serve 2 wildcard domains (and the 2 base domains as well).
4. For wildcard domains, dns challenge is used (with the help of [netcup DNS module](https://github.com/caddy-dns/netcup)).

Kind regards,
aanno

**Reply by aanno** [reply]
There is an ongoing discussion about this at https://github.com/nextcloud/all-in-one/discussions/5994 

Hence, don't reply here, but there!
